### PR TITLE
Cloud Metrics Beta

### PR DESCRIPTION
**Pull request by chakrabortyr**
## Description
As the above suggests, we are adding a new realm to track cloud data in XDMoD.

Example:
xdmod_resource_utilization_vs_vm_instantiation__2017-01-01_to_2017-12-31

## Motivation and Context
We need this to monitor utilization of cloud resources.

## Tests performed
Matching arithmetic results against what was in the SQL database. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: configuration/etl/etl_sql.d/cloud_eucalyptus/jobs_cloud_eucalyptus.json [hunk 1]
```diff
@@ -0,0 +1,203 @@
+{
```
Comment (eiffel777) [inline configuration/etl/etl_sql.d/cloud_eucalyptus/jobs_cloud_eucalyptus.json]: I don't think this file is supposed to be here. It seems like a duplicate of configuration/etl/etl.d/cloud_eucalyptus/jobs_cloud_eucalyptus.json

Comment (chakrabortyr) [reply configuration/etl/etl_sql.d/cloud_eucalyptus/jobs_cloud_eucalyptus.json]: Good catch, not sure how this ended up here. Removing.

File: classes/DataWarehouse/Query/Cloud/Raw.php [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        $this->addWhereCondition(
+            new \DataWarehouse\Query\Model\WhereCondition(
+                new \DataWarehouse\Query\Model\Field("end_time_ts"),
+                '<=',
+                new \DataWarehouse\Query\Model\Field("unix_timestamp('{$this->_end_date}')+86399")
+            )
+        );
+    }
+
+    public function exportJsonStore($limit = 20, $offset = 0)
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Raw.php]: Why is this the only class that includes `exportJsonStore`?

File: classes/DataWarehouse/Query/Cloud/GroupBys/GroupByYear.php [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+		$query->addField($value_field);
+
+		$query->addGroup($id_field);
+
+		$this->addOrder($query,$multi_group);
+	}
+	public function addOrder(\DataWarehouse\Query\Query &$query, $multi_group = false, $dir = 'asc', $prepend = false)
+	{
+		$orderField = new \DataWarehouse\Query\Model\OrderBy(new \DataWarehouse\Query\Model\TableField($query->getDataTable(), "year_id"), $dir, $this->getName());
+		if($prepend === true)
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/GroupBys/GroupByYear.php]: This isn't going to pass style checks. Braces should be on the same line, consider using `phpcbf`.

File: classes/DataWarehouse/Query/Cloud/Raw.php [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+            $columns[] = array("header" => ucwords(str_replace('_', ' ', $columnMeta['name'])), "width" => 100, "dataIndex" => $columnMeta['name'],
+              "sortable" => true, 'editable' => false, 'align' => 'left','xtype' => 'gridcolumn', 'locked' => false);
+        }
+
+        while($result = $statement->fetch(\PDO::FETCH_ASSOC, \PDO::FETCH_ORI_NEXT))
+        {
+            $records[] = $result;
+        }
+
+        $returnData = array(
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Raw.php]: Why set the variable and not just `return array(...`?

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            'avg_core_time',
+            'Average Core Hours Per VM',
+            'Hours',
+            0
+        );
+    }
+
+    public function getInfo()
+    {
+        return 'The average core time of virtual machines in a given period, in hours.<br/><i>Only approximate in aggregate view.</i><br/>
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php]: Should be "The average code hours used by virtual machines...", no?

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+            <b>Core Time</b>: The product of the number of cores reserved by a VM and its wall time.';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php]: Core Hours, Core Time is ambiguous.

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageCoresReservedStatistic.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+            'avg_num_cores',
+            'Cores Reserved: Average Per VM',
+            'Cores',
+            0
+        );
+    }
+
+    public function getInfo()
+    {
+        return 'The average number of cores reserved by virtual machines.<br/>.';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageCoresReservedStatistic.php]: Is this reserved or in use or something else? Would "consumed" be a better term?

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageWallHoursStatistic.php [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+            'avg_wallduration_hours',
+            'Average Wall Hours per VM',
+            'Hours',
+            0
+        );
+    }
+
+    public function getInfo()
+    {
+        return "The average time in hours in which a virtual machine was running.<br/> <i>Only approximate in aggregate view</i>.<br />
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageWallHoursStatistic.php]: More succinctly: "The average number of hours that a virtual machine was running."

File: classes/DataWarehouse/Query/Cloud/Statistics/CoreHoursStatistic.php [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+            'core_time',
+            'Core Hours: Total',
+            'Hours',
+            0
+        );
+    }
+
+    public function getInfo()
+    {
+        return 'The total core reservation time of virtual machines, in hours.<br/>';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/CoreHoursStatistic.php]: What is "core reservation time"?

File: classes/DataWarehouse/Query/Cloud/Statistics/MemoryReservationStatistic.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+            'memory_reserved',
+            'Memory Utilized Per Period',
+            'Bytes',
+            0
+        );
+    }
+
+    public function getInfo()
+    {
+        return 'The quotient of memory reserved by virtual machines over wall duration.<br/>';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/MemoryReservationStatistic.php]: Perhaps "consumed" rather than "reserved"?

File: classes/DataWarehouse/Query/Cloud/Statistics/VMCountStatistic.php [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+            'num_vms_ended',
+            'Number of VMs Ended',
+            'Number of VMs',
+            0
+        );
+    }
+
+    public function getInfo()
+    {
+        return 'The total number of ' . ORGANIZATION_NAME . ' VM instances that ended within the selected duration.<br/>';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/VMCountStatistic.php]: What is "ended"? Terminated, stopped, shelved, etc?

File: classes/DataWarehouse/Query/Cloud/Statistics/WeightedCoreWallStatistic.php [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+            'Cores Reserved: Weighted by Core Time',
+            'Cores',
+            0
+        );
+    }
+
+    public function getInfo()
+    {
+        return 'The number of cores reserved by virtual machines weighted by core time.<br/>
+                <b>Core Time</b>: The product of the number of cores reserved by a VM and its wall time.';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/WeightedCoreWallStatistic.php]: Hours...

File: configuration/etl/etl.d/jobs_cloud_common.json [hunk 13]
```diff
@@ -102,6 +102,32 @@
...
             }
+        },
+        {
+            "name": "memorybuckets",
+            "definition_file": "cloud_common/memory_buckets.json",
+            "description": "processorbuckets + data",
+            "endpoints": {
+                "source": {
+                    "type": "jsonfile",
+                    "name": "processorbuckets data",
```
Comment (smgallo) [inline configuration/etl/etl.d/jobs_cloud_common.json]: Should be "memory buckets data"

File: configuration/etl/etl_action_defs.d/cloud_common/cloud_metrics_aggregation.json [hunk 14]
```diff
@@ -0,0 +1,215 @@
+{
+    "#": "Aggregation of Cloud Data ingested from Eucalyptus",
```
Comment (smgallo) [inline configuration/etl/etl_action_defs.d/cloud_common/cloud_metrics_aggregation.json]: If this is from Euca then it should not be in the `cloud_common` directory.

File: configuration/etl/etl_action_defs.d/cloud_common/cloud_metrics_aggregation.json [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+        "table_definition": {
```
Comment (smgallo) [inline configuration/etl/etl_action_defs.d/cloud_common/cloud_metrics_aggregation.json]: Table definitions now go into `etl/etl_tables.d` and are named similar to "jobfact_by_.json" Use `$ref` to include them here.

File: configuration/etl/etl_action_defs.d/cloud_common/cloud_metrics_aggregation.json [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+            "name": "cloud_fact_by_",
```
Comment (smgallo) [inline configuration/etl/etl_action_defs.d/cloud_common/cloud_metrics_aggregation.json]: Use "cloudfact" rather than "cloud_fact" for consistency with other aggregate tables.

File: configuration/etl/etl_sql.d/cloud_common/post_ingest_final.sql [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    itt.start_time AS config_start,
+    UNIX_TIMESTAMP(consolidate.start_event_time) AS start_time_ts,
+    UNIX_TIMESTAMP(consolidate.end_event_time) AS end_time_ts,
+    UNIX_TIMESTAMP(consolidate.end_event_time) - UNIX_TIMESTAMP(consolidate.start_event_time) AS wallduration,
+    YEAR(consolidate.start_event_time) * 100000 + DAYOFYEAR(consolidate.start_event_time) AS start_day_id,
+    YEAR(consolidate.end_event_time) * 100000 + DAYOFYEAR(consolidate.end_event_time) AS end_day_id
+FROM (
+    -- Match up instance stop events that occur after a start event
+    SELECT
+		start_event.event_id,
```
Comment (smgallo) [inline configuration/etl/etl_sql.d/cloud_common/post_ingest_final.sql]: Use spaces not tabs.

File: configuration/etl/etl_sql.d/cloud_common/post_ingest_final.sql [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        start_event.start_event_type,
+        start_event.start_event_time,
+        end_event.end_event_type,
+        end_event.end_event_time,
+        start_event.instance_id,
+        start_event.resource_id
+    FROM (
+        -- Find all events where an instance is starting
+        SELECT
+			ev.event_id,
```
Comment (smgallo) [inline configuration/etl/etl_sql.d/cloud_common/post_ingest_final.sql]: Use spaces not tabs.

File: configuration/etl/etl_sql.d/cloud_eucalyptus/jobs_cloud_eucalyptus.json [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+            "class": "DatabaseIngestor",
+            "definition_file": "cloud_eucalyptus/instance.def.json"
+        },
+        {
+            "name": "EucalyptusCloudVolumeAssetIngestor",
+            "description": "Eucalyptus cloud attachable volume (block device) data from raw logs",
+            "class": "DatabaseIngestor",
+            "definition_file": "cloud_eucalyptus/volume.def.json"
+        },
+        {
```
Comment (smgallo) [inline configuration/etl/etl_sql.d/cloud_eucalyptus/jobs_cloud_eucalyptus.json]: If this is a post-ingest SQL, it should be after ingestion not in the middle of it.

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageWallHoursStatistic.php [hunk 20]
```diff
@@ -24,7 +24,7 @@ public function __construct($query_instance)
 
     public function getInfo()
     {
-        return "The average time in hours in which an instantiated virtual machine was running.<br/> <i>Only approximate in aggregate view</i>.<br />
+        return "The average time a virtual machine was running, in hours.<br/> <i>Only approximate in aggregate view</i>.<br />
             <i>Wall Time:</i> Wall time is defined as the linear duration between the start and end times of discrete virtual machine runs.";
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageWallHoursStatistic.php]: Is this actually the sum of the discrete runs?

Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageWallHoursStatistic.php]: Is this actually the sum of the discrete runs?

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php [hunk 21]
```diff
@@ -23,7 +23,7 @@ public function __construct($query_instance = null)
 
     public function getInfo()
     {
-        return 'The average core time of instantiated virtual machines in a given period, in hours.<br/><i>Only approximate in aggregate view.</i><br/>
+        return 'The average core time of running virtual machines, in hours.<br/><i>Only approximate in aggregate view.</i><br/>
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php]: Core Time is ambiguous. Also "reservations" have other meaning in HPC. Why is it only approximate?

How about: 'The average core hours consumed by running virtual machines
Only approximate in aggregate view.
Core Hours: The product of the number of cores consumed by a VM and its wall time, in hours.'

File: classes/DataWarehouse/Query/Cloud/Statistics/CoreHoursStatistic.php [hunk 22]
```diff
@@ -22,6 +22,6 @@ public function __construct($query_instance = null)
 
     public function getInfo()
     {
-        return 'The total core reservation time of virtual machines, in hours.<br/>';
+        return 'The total core reservation time of running virtual machines, in hours.<br/>';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/CoreHoursStatistic.php]: "The number of core hours consumed by running virtual machines"

File: classes/DataWarehouse/Query/Cloud/Statistics/CoresReservedStatistic.php [hunk 23]
```diff
@@ -22,6 +22,6 @@ public function __construct($query_instance = null)
 
     public function getInfo()
     {
-        return 'The total number of cores reserved by virtual machines.<br/>';
+        return 'The total number of cores reserved by running virtual machines.<br/>';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/CoresReservedStatistic.php]: Lets find another word other than reserved - how about "allocated"? "The total number of cores allocated to running virtual machines."

File: classes/DataWarehouse/Query/Cloud/Statistics/DiskReservationStatistic.php [hunk 24]
```diff
@@ -22,6 +22,6 @@ public function __construct($query_instance = null)
 
     public function getInfo()
     {
-        return 'The quotient of disk space reserved by virtual machines over wall duration.<br/>';
+        return 'The quotient of disk space reserved by running virtual machines over wall duration.<br/>';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/DiskReservationStatistic.php]: "The amount of disk space allocated to running virtual machines." Is this divided by the wallduration?

File: classes/DataWarehouse/Query/Cloud/Statistics/MemoryReservationStatistic.php [hunk 25]
```diff
@@ -22,6 +22,6 @@ public function __construct($query_instance = null)
 
     public function getInfo()
     {
-        return 'The quotient of memory reserved by virtual machines over wall duration.<br/>';
+        return 'The quotient of memory reserved by running virtual machines over wall duration.<br/>';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/MemoryReservationStatistic.php]: Lets use "amount" instead of "quotient".

File: classes/DataWarehouse/Query/Cloud/Statistics/VMCountStatistic.php [hunk 26]
```diff
@@ -22,6 +22,6 @@ public function __construct($query_instance = null)
 
     public function getInfo()
     {
-        return 'The total number of VM instances that were shelved or terminated.<br/>';
+        return 'The total number of running virtual machines that were shelved or terminated.<br/>';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/VMCountStatistic.php]: Be consistent either add "on a cloud resource" here or remove it from StartedVMCountStatistic.php

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageWallHoursStatistic.php (hunk 20 above)
Comment (chakrabortyr) [reply classes/DataWarehouse/Query/Cloud/Statistics/AverageWallHoursStatistic.php]: Yes it is. I can maybe reword this to clarify

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php [hunk 27]
```diff
@@ -24,6 +24,6 @@ public function __construct($query_instance = null)
     public function getInfo()
     {
         return 'The average core hours consumed by running virtual machines.<br/><i>Only approximate in aggregate view.</i><br/>
-            <b>Core Time</b>: The product of the number of cores consumed by a VM and its wall time, in hours.';
+            <b>Core Time</b>: The product of the number of cores allocated to a VM and its wall time, in hours.';
```
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php]: Change "Core Time" to "Core Hours"

File: classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php (hunk 21 above)
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/AverageCoreHoursStatistic.php]: Lets change "core time" to "core hours" here as well.

File: classes/DataWarehouse/Query/Cloud/Statistics/MemoryReservationStatistic.php (hunk 25 above)
Comment (smgallo) [inline classes/DataWarehouse/Query/Cloud/Statistics/MemoryReservationStatistic.php]: Lets use "amount" instead of "quotient".

**Comment by jsperhac** [comment]
**Feedback on the Description text**

I'm attaching, for each metric, the description pane contents (pulled off your cloud metrics dev instance, rudy-dhmnm2.ccr.xdmod.org) and any questions/remarks I had about that metric's descriptions. I have some of the same remarks that Steve mentions above.

My primary remark is that "core time" and "core hours" appear to be calculated the same way, and should accordingly be labeled and defined consistently throughout all the metrics. (I like the definition you use in Average core hours per vm, but I would label it "Core Hours" there.)

The dimension descriptions (not included here) look ok. Correct the spelling of the word 'amount' ("amound") in VM Size: Memory dimension. Just let me know if you need clarification.

[cloud-metrics-helptext.txt](https://github.com/ubccr/xdmod/files/2031111/cloud-metrics-helptext.txt)

**Comment by chakrabortyr** [comment]
Thanks @smgallo and @jsperhac. Will incorporate feedback and get the descriptions consistent!

File: open_xdmod/modules/xdmod/tests/lib/Xdmod/ConfigTest.php [hunk 28]
```diff
@@ -267,8 +267,8 @@ public function testFilteringWithNonAssociativeArrays()
 
     public function moduleSectionProvider()
     {
-        $rolesExpected = Json::loadFile(__DIR__ . self::TEST_ARTIFACT_OUTPUT_PATH . DIRECTORY_SEPARATOR . 'roles-groupbywaittime.json');
-        $datawarehouseExpected = Json::loadFile(__DIR__ . self::TEST_ARTIFACT_OUTPUT_PATH . DIRECTORY_SEPARATOR . 'datawarehouse-groupbywaittime.json');
+        $rolesExpected = Json::loadFile(__DIR__ . self::TEST_ARTIFACT_OUTPUT_PATH . DIRECTORY_SEPARATOR . 'roles-cloud.json');
```
Comment (jpwhite4) [inline open_xdmod/modules/xdmod/tests/lib/Xdmod/ConfigTest.php]: Now that the test artifacts have been merged you can update the artifact directly in this pull request rather than creating a new file.

**Comment by plessbd** [comment]
I would like to see some anonymized data similar to what we have for the jobs realm put into the docker with some tests against it so we can have regression tests for this data, rigth now all we have is configuration tests...

**Comment by smgallo** [comment]
@plessbd Agreed, but given our short timeframe that can be done in a subsequent PR.

**Comment by plessbd** [comment]
can we get some of these commits squashed... It will be easier to review
some of them seem to be switching back and forth on the same things and not really useful

ie.
```
fa195352af160ed2c289068355db26cb42a92720
f213809d4c639415cb0723de4e748240617feaee
```

File: classes/DataWarehouse/Query/Cloud/GroupBy.php [hunk 29]
```diff
@@ -0,0 +1,45 @@
...
+    public function __construct($name, array $additional_permitted_parameters = array(), $possible_values_query = null)
+    {
+        $permitted_paramters = array_unique(
+            array_merge(
+                array_keys(Aggregate::getRegisteredStatistics()),
+                $additional_permitted_parameters
+            )
+        );
+
+        parent::__construct($name, $permitted_paramters, $possible_values_query);
```
Comment (plessbd) [inline classes/DataWarehouse/Query/Cloud/GroupBy.php]: spelling fail `permitted_paramters` should probably be `permitted_parameters`

Also, is there a need for the extra variable at all when you can just do
```php
    parent::__construct(
        $name,
        array_unique(
            array_merge(
                array_keys(Aggregate::getRegisteredStatistics()),
                $additional_permitted_parameters
            )
        ),
        $possible_values_query
    );
```

File: configuration/etl/etl_data.d/cloud_common/memory_buckets.json [hunk 30]
```diff
@@ -0,0 +1,12 @@
+[
```
[the remark this answers is not in the document]
Comment (plessbd) [reply configuration/etl/etl_data.d/cloud_common/memory_buckets.json]: seriously who does 30, 60,  gb, the CS degrees must be revoked!
Also, if the above atlassian site doesnt work ( I got a corrupt network protocol error) you can use: https://jetstream-cloud.org/tech-specs/cloud-services.php

[the remark this answers is not in the document]
Comment (smgallo) [reply configuration/etl/etl_data.d/cloud_common/memory_buckets.json]: I suspect that it might have something to do with the SU calculations, but they do look odd.

File: classes/DataWarehouse/Query/Cloud/GroupBy.php (hunk 29 above)
Comment (smgallo) [reply classes/DataWarehouse/Query/Cloud/GroupBy.php]: This spelling should still be fixed

File: configuration/etl/etl_data.d/cloud_common/processor_buckets.json [hunk 31]
```diff
@@ -0,0 +1,9 @@
+[
+    ["id", "min_processors", "max_processors", "description"],
+    [1,       1,           1, "1"],
+    [2,       2,           4, "2 - 4"],
+    [3,       5,           8, "5 - 8"],
+    [4,       9,           16, "9 - 16"],
+    [5,       17,          32, "17 - 32"],
+    [6,       33,         96, "> 32"],
```
Comment (smgallo) [inline configuration/etl/etl_data.d/cloud_common/processor_buckets.json]: Remove trailing comma.

File: classes/ETL/Ingestor/StateReconstructorTransformIngestor.php [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+            return array();
+        }
+
+        $transformedRecord = array();
+
+        if ($this->instance_state['instance_id'] !== $srcRecord['instance_id']) {
+            $transformedRecord[] = $this->instance_state;
+            $this->initInstance($srcRecord);
+        }
+        else if  (in_array($srcRecord['event_type_id'], $this->start_event_ids)) {
```
Comment (smgallo) [inline classes/ETL/Ingestor/StateReconstructorTransformIngestor.php]: Use `elseif`

File: classes/ETL/Ingestor/StateReconstructorTransformIngestor.php [hunk 33]
```diff
@@ -0,0 +1,115 @@
...
+            $this->updateInstance($srcRecord);
+        }
+        else if (in_array($srcRecord['event_type_id'], $this->stop_event_ids)) {
```
Comment (smgallo) [inline classes/ETL/Ingestor/StateReconstructorTransformIngestor.php]: Use `elseif`

File: open_xdmod/modules/xdmod/tests/lib/Xdmod/ConfigTest.php (hunk 28 above)
Comment (smgallo) [reply open_xdmod/modules/xdmod/tests/lib/Xdmod/ConfigTest.php]: You can just copy `xdmod/acls/output/roles-cloud.json` from the test artifacts repo directly into `tests/artifacts/xdmod-test-artifacts` in this repo.

File: classes/DataWarehouse/Query/Cloud/Raw.php [hunk 34]
```diff
@@ -0,0 +1,201 @@
+<?php
+namespace DataWarehouse\Query\Cloud;
```
Comment (jpwhite4) [inline classes/DataWarehouse/Query/Cloud/Raw.php]: Remove this file

File: configuration/etl/etl_data.d/cloud_common/memory_buckets.json [hunk 35]
```diff
@@ -0,0 +1,11 @@
...
+    ["id", "min_memory", "max_memory", "description"],
+    [1, 1, 1024, "1GB"],
```
Comment (jpwhite4) [inline configuration/etl/etl_data.d/cloud_common/memory_buckets.json]: No gap

File: configuration/etl/etl_sql.d/cloud_common/post_ingest_final.sql [hunk 36]
```diff
@@ -0,0 +1,69 @@
+DROP TABLE IF EXISTS cloud_events_transient;
```
Comment (jpwhite4) [inline configuration/etl/etl_sql.d/cloud_common/post_ingest_final.sql]: remove this

Comment (jpwhite4) [reply configuration/etl/etl_sql.d/cloud_common/post_ingest_final.sql]: remove and delete are synonymous for the purpose of these comments on this pull request.

File: configuration/etl/etl_data.d/cloud_common/processor_buckets.json [hunk 37]
```diff
@@ -0,0 +1,9 @@
+[
+    ["id", "min_processors", "max_processors", "description"],
+    [1,       1,           1, "1"],
+    [2,       2,           4, "2 - 4"],
+    [3,       5,           8, "5 - 8"],
+    [4,       9,           16, "9 - 16"],
+    [5,       17,          32, "17 - 32"],
+    [6,       33,         96, "> 32"]
```
Comment (jpwhite4) [inline configuration/etl/etl_data.d/cloud_common/processor_buckets.json]: make the 96 max(int)

File: configuration/etl/etl_action_defs.d/cloud_common/cloud_metrics_aggregation.json [hunk 38]
```diff
@@ -0,0 +1,91 @@
...
+                "memorybucket_id": "(SELECT id FROM ${SOURCE_SCHEMA}.memory_buckets mb WHERE cet.memory_mb BETWEEN mb.min_memory AND mb.max_memory)",
+                "instance_type_id": "cet.instance_type_id",
+                "configuration": "cet.instance_type",
+                "num_cores": "SUM(cet.num_cores)",
+                "core_time": "COALESCE(SUM(cet.num_cores * ${wallduration_case_statement}), 0)",
+                "memory_reserved": "COALESCE(SUM((cet.memory_mb * POW(1024, 2)) * ${wallduration_case_statement}), 0)",
+                "disk_reserved": "COALESCE(SUM((cet.disk_gb * POW(1024, 3)) * ${wallduration_case_statement}), 0)",
+                "memory_mb": "SUM(cet.memory_mb)",
+                "disk_gb": "SUM(cet.disk_gb)",
+                "num_vms_started": "SUM(IF(cet.start_day_id BETWEEN ${:PERIOD_START_DAY_ID} AND ${:PERIOD_END_DAY_ID} AND cet.start_event_type_id IN ('2', '8'), 1, 0))",
```
Comment (jpwhite4) [inline configuration/etl/etl_action_defs.d/cloud_common/cloud_metrics_aggregation.json]: do not restrict by event_type_id
